src/engine/qcommon/q_math.cpp

```diff
@@ -549,8 +549,14 @@ float ProjectPointOntoRectangleOutwards( vec2_t out, const vec2_t point, const v
 	dsign[ 0 ] = ( dir[ 0 ] < 0.0f );
 	dsign[ 1 ] = ( dir[ 1 ] < 0.0f );
 
-	t = ( bounds[ 1 - dsign[ 0 ] ][ 0 ] - point[ 0 ] ) / dir[ 0 ];
-	ty = ( bounds[ 1 - dsign[ 1 ] ][ 1 ] - point[ 1 ] ) / dir[ 1 ];
+	/* Avoids division-by-zero.
```

Comment:
I don't like this one. It looks more like sweeping bugs under the rug rather than fixing them

Reply:
All three callers of this code can feed a legit 0.

Reply:
Maybe the conceptual mistake in such context is to assume a float can be equal to an integer to begin with, not just `0`.

Reply:
The three callers of `ProjectPointOntoRectangleOutwards` are things drawing beacons:

- `src/cgame/cg_beacon.cpp`, `DrawBeacon`:

```c++
static void DrawBeacon( cbeacon_t *b )
{
// …
	// project onto screen
	front = CG_WorldToScreen( b->origin, b->pos, b->pos + 1);

	// CG_WorldToScreen flips the result if behind so correct it
	if( !front )
	{
		b->pos[ 0 ] = 640.0 - b->pos[ 0 ],
		b->pos[ 1 ] = 480.0 - b->pos[ 1 ];
	}

	// virtual 640x480 to real
	b->pos[ 0 ] *= cgs.glconfig.vidWidth / 640.0;
	b->pos[ 1 ] *= cgs.glconfig.vidHeight / 480.0;

	// clamp to edges
	if( !front ||
	    b->pos[ 0 ] < cgs.bc.hudRect[0][0] + b->size/2 ||
	    b->pos[ 0 ] > cgs.bc.hudRect[1][0] - b->size/2 ||
	    b->pos[ 1 ] < cgs.bc.hudRect[0][1] + b->size/2 ||
	    b->pos[ 1 ] > cgs.bc.hudRect[1][1] - b->size/2 )
	{
		vec2_t screen[ 2 ];
		Vector2Set( screen[ 0 ], cgs.bc.hudRect[0][0] + b->size/2,
		                         cgs.bc.hudRect[0][1] + b->size/2 );
		Vector2Set( screen[ 1 ], cgs.bc.hudRect[1][0] - b->size/2,
		                         cgs.bc.hudRect[1][1] - b->size/2 );
		Vector2Subtract( b->pos, cgs.bc.hudCenter, b->clamp_dir );
		ProjectPointOntoRectangleOutwards( b->pos, cgs.bc.hudCenter, b->clamp_dir, (const vec2_t*)screen );
		b->clamped = true;
	}
	else
		b->clamped = false;
}
```

- `src/cgame/cg_draw.cpp`, `CG_DrawBeacon`:

```c++
static void CG_DrawBeacon( cbeacon_t *b )
{
// …
	if( b->type == BCT_TIMER )
	{
		int num;

		num = ( BEACON_TIMER_TIME + b->ctime - cg.time ) / 100;

		if( num > 0 )
		{
			float h, tw;
			const char *p;
			vec2_t pos, dir, rect[ 2 ];
			int i, l, frame;

			h = b->size * 0.4f;
			p = va( "%d", num );
			l = strlen( p );
			tw = h * l;

			if( !b->clamped )
			{
				pos[ 0 ] = b->pos[ 0 ];
				pos[ 1 ] = b->pos[ 1 ] + b->size/2 + h/2;
			}
			else
			{
				rect[ 0 ][ 0 ] = b->pos[ 0 ] - b->size/2 - tw/2;
				rect[ 1 ][ 0 ] = b->pos[ 0 ] + b->size/2 + tw/2;
				rect[ 0 ][ 1 ] = b->pos[ 1 ] - b->size/2 - h/2;
				rect[ 1 ][ 1 ] = b->pos[ 1 ] + b->size/2 + h/2;

				for( i = 0; i < 2; i++ )
					dir[ i ] = - b->clamp_dir[ i ];

				ProjectPointOntoRectangleOutwards( pos, b->pos, dir, (const vec2_t*)rect );
			}

			pos[ 0 ] -= tw/2;
			pos[ 1 ] -= h/2;

			for( i = 0; i < l; i++ )
			{
				if( p[ i ] >= '0' && p[ i ] <= '9' )
					frame = p[ i ] - '0';
				else if( p[ i ] == '-' )
					frame = STAT_MINUS;
				else
					frame = -1;

				if( frame != -1 )
					trap_R_DrawStretchPic( pos[ 0 ], pos[ 1 ], h, h, 0, 0, 1, 1, cgs.media.numberShaders[ frame ] );

				pos[ 0 ] += h;
			}
		}
	}

	trap_R_ClearColor();
}
```

- `src/cgame/cg_minimap.cpp`, `CG_MinimapDrawBeacon`:

```c++
static void CG_MinimapDrawBeacon( const cbeacon_t *b, float size, const vec2_t center, const vec2_t *bounds )
{
	vec2_t offset, pos2d, dir;
	bool clamped;

	size *= b->scale;

	CG_WorldToMinimap( b->origin, offset );
	pos2d[ 0 ] = - size/2 + offset[ 0 ];
	pos2d[ 1 ] = - size/2 + offset[ 1 ];

	if( pos2d[ 0 ] < bounds[ 0 ][ 0 ] ||
	    pos2d[ 0 ] > bounds[ 1 ][ 0 ] ||
	    pos2d[ 1 ] < bounds[ 0 ][ 1 ] ||
	    pos2d[ 1 ] > bounds[ 1 ][ 1 ] )
	{
		clamped = true;

		if( b->type == BCT_TAG )
			return;

		Vector2Subtract( pos2d, center, dir );
		ProjectPointOntoRectangleOutwards( pos2d, center, dir, bounds );
	}
	else
	{
		clamped = false;
	}

	Color::Color color = b->color;
	color.SetAlpha( cgs.bc.minimapAlpha );
	trap_R_SetColor( color );

	trap_R_DrawRotatedPic( pos2d[ 0 ], pos2d[ 1 ], size, size, 0.0f, 0.0f, 1.0f, 1.0f, CG_BeaconIcon( b ), 0.0f );
	if( b->flags & EF_BC_DYING )
		trap_R_DrawStretchPic( pos2d[ 0 ] - size/2 * 0.3f,
		                       pos2d[ 1 ] - size/2 * 0.3f,
		                       size * 1.3, size * 1.3f,
		                       0, 0, 1, 1,
		                       cgs.media.beaconNoTarget );
	if( clamped )
		trap_R_DrawRotatedPic( pos2d[ 0 ] - size * 0.25f,
		                       pos2d[ 1 ] - size * 0.25f,
		                       size * 1.5, size * 1.5f,
		                       0.0f, 0.0f, 1.0f, 1.0f,
		                       cgs.media.beaconIconArrow,
		                       270.0f - atan2f( dir[ 1 ], dir[ 0 ] ) * 180 / M_PI );
}
```